src/BidiStreamingResponse.php

```diff
@@ -0,0 +1,174 @@
+     *
+     * @param mixed $request The request to write
+     * @throws ValidationException
+     */
+    public function write($request)
+    {
+        if ($this->writesClosed) {
+            throw new ValidationException("Cannot call write() after calling closeWrite().");
+        }
+        if ($this->isComplete) {
+            throw new ValidationException("Cannot call write() after streaming call is complete.");
+        }
+        $this->call->write($request);
+    }
+
+    /**
+     * Write all requests in $requests.
+     *
+     * @param mixed[] $requests An Iterable of request objects to write to the server
+     *
+     * @throws ValidationException
+     * @throws ApiException
+     */
+    public function writeAll($requests = [])
+    {
+        foreach ($requests as $request) {
+            $this->write($request);
+        }
+    }
+
+    /**
+     * Inform the server that no more requests will be written. The write() function cannot be
+     * called after closeWrite() is called.
+     */
+    public function closeWrite()
+    {
+        if ($this->isComplete) {
+            throw new ValidationException(
+                "Cannot call closeWrite() after streaming call is complete."
+            );
+        }
+        if (!$this->writesClosed) {
+            $this->call->writesDone();
+            $this->writesClosed = true;
+        }
+    }
+
+    /**
+     * Read the next response from the server. Returns null if the streaming call completed
+     * successfully. Throws an ApiException if the streaming call failed.
+     *
+     * @return mixed
+     * @throws ValidationException
+     * @throws ApiException
+     */
+    public function read()
+    {
+        if ($this->isComplete) {
+            throw new ValidationException("Cannot call read() after streaming call is complete.");
+        }
+        $resourcesField = $this->resourcesField;
+        if (!is_null($resourcesField)) {
+            if (count($this->pendingResources) === 0) {
+                $response = $this->call->read();
+                if (!is_null($response)) {
+                    $this->pendingResources = array_reverse($response->$resourcesField());
+                }
+            }
+            $result = array_pop($this->pendingResources);
+        } else {
+            $result = $this->call->read();
+        }
+        if (is_null($result)) {
+            $status = $this->call->getStatus();
+            $this->isComplete = true;
+            if (!($status->code == Grpc\STATUS_OK)) {
+                throw new ApiException($status->details, $status->code);
+            }
+        }
+        return $result;
+    }
+
+    /**
+     * Call closeWrite(), and read all responses from the server, until the streaming call is
+     * completed. Throws an ApiException if the streaming call failed.
+     *
+     * @return \Generator|mixed[]
+     * @throws ValidationException
+     * @throws ApiException
+     */
+    public function closeAndReadAll()
```

Comment:
This name is confusing because it doesn't make sense that you can read anything after closing. I see that it is actually closing the write specifically. Why do we need this combination - can't the use call `closeWrite()` and then call a `readAll()` function after?